File: app/src/main/java/weatherapp/tittojose/me/weatherapp/ui/view/HomeScreenActivity.java
package weatherapp.tittojose.me.weatherapp.ui.view;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.constraint.ConstraintLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import weatherapp.tittojose.me.weatherapp.R;
import weatherapp.tittojose.me.weatherapp.Utils;
import weatherapp.tittojose.me.weatherapp.model.APIClient;
import weatherapp.tittojose.me.weatherapp.model.WeatherAPI;
import weatherapp.tittojose.me.weatherapp.model.pojo.WeatherModel;
import weatherapp.tittojose.me.weatherapp.repository.WeatherRepository;
import weatherapp.tittojose.me.weatherapp.repository.WeatherRepositoryContract;
import weatherapp.tittojose.me.weatherapp.ui.presenter.HomeScreenPresenter;

public class HomeScreenActivity extends AppCompatActivity implements HomeScreen {

    private WeatherAPI weatherAPIClient;

    @BindView(R.id.txtCurrentWeather)
    TextView currentTemperatureTextView;

    @BindView(R.id.rvForecast)
    RecyclerView forecastRecyclerview;

    @BindView(R.id.txtLocation)
    TextView locationTextView;

    @BindView(R.id.btnRetry)
    Button retryButton;

    @BindView(R.id.imgViewLoading)
    ImageView loadingImageView;

    @BindView(R.id.layoutLoading)
    ViewGroup loadingLayout;

    @BindView(R.id.layoutWeather)
    ConstraintLayout weatherLayout;

    @BindView(R.id.layoutError)
    ViewGroup errorLayout;
    private Animation startRotateAnimation;
    private HomeScreenPresenter homeScreenPresenter;
    private Animation bottomUpAnimation;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home_screen);
        ButterKnife.bind(this);
        initializeAnimationObjects();

        weatherAPIClient = APIClient.getWeatherAPIClient();

        WeatherRepositoryContract weatherRepositoryContract = new WeatherRepository(weatherAPIClient);

        homeScreenPresenter = new HomeScreenPresenter(this, weatherRepositoryContract);
        homeScreenPresenter.loadWeatherData();
        retryButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onRetryButtonClicked();
            }
        });

    }

    private void initializeAnimationObjects() {
        startRotateAnimation = AnimationUtils.loadAnimation(getApplicationContext(), R.anim.anim_rotate);
        bottomUpAnimation = AnimationUtils.loadAnimation(getApplicationContext(), R.anim.anim_bottomsup);
    }

    @Override
    public void onWeatherDataLoadSuccess(WeatherModel weatherModel) {
        weatherLayout.setVisibility(View.VISIBLE);
        setCurrentWeatherData(weatherModel);
        setForecastWeatherData(weatherModel);
    }

    @Override
    public void onWeatherDataLoadFailed() {
        errorLayout.setVisibility(View.VISIBLE);
    }

    @Override
    public void showLoading() {
        loadingLayout.setVisibility(View.VISIBLE);
        loadingImageView.startAnimation(startRotateAnimation);
    }

    @Override
    public void hideLoading() {
        loadingImageView.clearAnimation();
        loadingLayout.setVisibility(View.GONE);
    }

    @Override
    public void onRetryButtonClicked() {
        homeScreenPresenter.retryLoadWeatherData();
    }


    private void setForecastWeatherData(WeatherModel weatherModel) {
        forecastRecyclerview.setAdapter(new ForecastRecyclerAdapter(weatherModel.getForecast()));
        forecastRecyclerview.setAnimation(bottomUpAnimation);
    }

    private void setCurrentWeatherData(WeatherModel weatherData) {
        currentTemperatureTextView.setText(Utils.getDegreeCelsiusData(weatherData.getCurrent().getTempC()));
        locationTextView.setText(weatherData.getLocation().getLocationName());
    }
}

File: app/src/main/java/weatherapp/tittojose/me/weatherapp/view/HomeScreenActivity.kt
package weatherapp.tittojose.me.weatherapp.view

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.view.View
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import kotlinx.android.synthetic.main.layout_error.*
import kotlinx.android.synthetic.main.layout_loading.*
import kotlinx.android.synthetic.main.layout_weather.*
import weatherapp.tittojose.me.weatherapp.R
import weatherapp.tittojose.me.weatherapp.model.APIClient
import weatherapp.tittojose.me.weatherapp.model.pojo.WeatherModel
import weatherapp.tittojose.me.weatherapp.presenter.HomeScreenPresenter
import kotlin.math.roundToInt

class HomeScreenActivity : AppCompatActivity(), HomeScreen {

	private var startRotateAnimation: Animation? = null
	private var bottomUpAnimation: Animation? = null
	private lateinit var homeScreenPresenter: HomeScreenPresenter

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_home_screen)
		initializeAnimationObjects()
		btnRetry.setOnClickListener { onRetryButtonClicked() }

		homeScreenPresenter = HomeScreenPresenter(this, APIClient.weatherAPIClient)
		homeScreenPresenter.loadWeatherData()
	}

	override fun onWeatherDataLoadSuccess(weatherModel: WeatherModel) {
		layoutWeather.visibility = View.VISIBLE
		setCurrentWeatherData(weatherModel)
	}

	override fun onWeatherDataLoadFailed() {
		layoutError.visibility = View.VISIBLE
	}

	override fun showLoading() {
		layoutLoading.visibility = View.VISIBLE
		imgViewLoading.startAnimation(startRotateAnimation)
	}

	override fun hideLoading() {
		imgViewLoading.clearAnimation()
		layoutLoading.visibility = View.GONE
	}

	override fun onRetryButtonClicked() {
		homeScreenPresenter.retryLoadWeatherData()
	}

	override fun onDestroy() {
		homeScreenPresenter.onDestroy()
		super.onDestroy()
	}

	private fun initializeAnimationObjects() {
		startRotateAnimation = AnimationUtils.loadAnimation(applicationContext, R.anim.anim_rotate)
		bottomUpAnimation = AnimationUtils.loadAnimation(applicationContext, R.anim.anim_bottomsup)
	}

	private fun setCurrentWeatherData(weatherData: WeatherModel) {
		txtCurrentWeather.text = formatDegreeCelsiusData(weatherData.current.tempC)
		txtLocation.text = weatherData.location.locationName
	}

	private fun formatDegreeCelsiusData(tempC: Float?): String {
		val tempInt = (tempC!!).roundToInt()
		return String.format("%s%s", tempInt, "\u2103")
	}
}